lib/wallet/migrations.js

```diff
@@ -0,0 +1,301 @@
+/*!
+ * wallet/migrations.js - blockchain data migrations for hsd
+ * Copyright (c) 2021, Nodari Chkuaselidze (MIT License).
+ * https://github.com/handshake-org/hsd
+ */
+
+'use strict';
+
+const assert = require('bsert');
+const Logger = require('blgr');
+const Network = require('../protocol/network');
+const {
+  AbstractMigration,
+  MigrationResult,
+  Migrations,
+  types,
+  oldLayout
+} = require('../migrations/migrations');
+const MigrationState = require('../migrations/state');
+const layouts = require('./layout');
+const layout = layouts.wdb;
+
+/**
+ * Switch to new migrations layout.
+ */
+
+class MigrateMigrations extends AbstractMigration {
+  /**
+   * Create migrations migration.
+   * @param {ChainMigrations} options
+   */
+
+  constructor(options) {
+    super(options);
+
+    this.options = options;
+    this.logger = options.logger.context('wallet-migrations-migrate');
+    this.db = options.db;
+    this.ldb = options.ldb;
+    this.layout = options.layout;
+  }
+
+  async check() {
+    return types.MIGRATE;
+  }
+
+  /**
+   * Actual migration
+   * @param {Batch} b
+   * @returns {Promise}
+   */
+
+  async migrate(b) {
+    this.logger.info('Migrating migrations..');
+    const state = new MigrationState();
+    state.nextMigration = 1;
+
+    if (await this.ldb.get(oldLayout.M.encode(0))) {
+      b.del(oldLayout.M.encode(0));
```

Comment:
What are the consequences of this for backwards compatibility? Like if the user for whatever reason upgrades, migrates, then tries to downgrade? I think we can prevent that case with good docs but just curious. (Blockstore wlll probably be hsd v3.0.0 anyway so broken compatibility should be clear)

Reply:
This is when you are upgrading from the latest version: You have already migrated change addresses. This part makes sure we clean up oldLayout entries as well ass make sure we don't rerun the change address migration. It's similar thing to chain/migrations - there I iterate over oldEntries because there might be 0 and 1. (Could be done similar way tbh)

Reply:
ok got it :+1: